File: ROID-Forum-Web-Client/Convert Client Code/GMGWebsite/web/Website/Sections/CodingSection/ThreadController.dart
library CodingThreadController.dart;
import 'ThreadInfo.dart';
import 'CodingSection.dart';
import 'dart:html';
import 'CommentInfo.dart';
import 'package:animation/animation.dart';

class ThreadController {
  List<ThreadInfo> threads;
  CodingSection sectionController;
  DivElement mainView;
  DivElement headerView;
  DivElement fullView;
  ButtonElement newPostButton, backButton;
  int viewingPosition = 0;
  ThreadInfo viewingThread = null;
  ThreadController(CodingSection c) {
    sectionController = c;
    threads = new List();
    mainView = new DivElement();
    headerView = new DivElement();
    SpanElement s = new SpanElement();
    s.text = "The Programmer's Corner";
    s.style
    ..position = "absolute"
    ..width = "100%"
    ..textAlign = "center"
    ..top = "14px"
    ..fontSize = "25px"
    ..color = "white";
    headerView.append(s);
    fullView = new DivElement();
    makeNewPostButtonComponent();
    makeBackButtonComponent();
    mainView.append(headerView);
  }
  void loggedInEvent() {
    headerView.append(newPostButton);
    for (int i = 0; i < threads.length; i++) {
      ThreadInfo t = threads[i];
      t.fullView.main.append(t.fullView.commentButtonAndBox);
    }
  }
  void loggedOutEvent() {
    newPostButton.remove();
    for (int i = 0; i < threads.length; i++) {
      ThreadInfo t = threads[i];
      t.fullView.commentButtonAndBox.remove();
    }
  }
  void restoreToDefaultState() {
    fullView.remove();
    fullView.children.clear();
    mainView.append(headerView);
    backButton.remove();
    viewingThread = null;
  }
  void threadClicked(ThreadInfo thread) {
    headerView.remove();
    mainView.append(fullView);
    fullView.append(thread.fullView.getDiv());
    mainView.append(backButton);
    viewingThread = thread;
  }
  void backButtonClicked() {
    restoreToDefaultState();
  }
  void updateThreadPositions() {
    for (int i = 0; i < threads.length; i++) {
      ThreadInfo t = threads[i];
      //t.headerView.getDiv().style.top = "${i*85+60}px";
      animate(t.headerView.getDiv(), duration: 250, properties: {
        'top': "${i*85+60}px"
      });
    }
  }
  ThreadInfo getThread(int id) {
    for (ThreadInfo t in threads) {
      if (t.getID() == id) return t;
    }
    return null;
  }
  void showThread(int threadID) {
    ThreadInfo thread = getThread(threadID);
    if (thread != null) {
      threadClicked(thread);
    }
  }
  //Actions
  void clearAllThreads() {
    for (int i = 0; i < threads.length; i++) {
      ThreadInfo t = threads[i];
      t.headerView.getDiv().remove();
      t.fullView.getDiv().remove();
    }
    threads.clear();
  }
  void addThread(Map threadMap) {
    int id = threadMap["ID"];
    if (getThread(id) == null) {
      ThreadInfo thread = new ThreadInfo(this);
      thread.setID(id);
      threads.add(thread);

      thread.setOwner(threadMap["Owner"]);
      if (threadMap["AvatarURL"] != null) {
        thread.setAvatarURL(threadMap["AvatarURL"]);
      }
      thread.setTitle(threadMap["Title"]);
      thread.setDescription(threadMap["Description"]);

      List<Map> commentArray = threadMap["Comments"];
      for (int i = 0; i < commentArray.length; i++) {
        Map c = commentArray[i];
        CommentInfo cf = new CommentInfo(thread, this);
        thread.addComment(cf);
        print("Loaded comment with threadID: ${c['ThreadID']} and commentID: ${c['CommentID']}");
        cf.setThreadID(c['ThreadID']);
        cf.setCommentID(c['CommentID']);
        cf.setComment(c['Comment']);
        cf.setOwner(c['Owner']);
        if (c["AvatarURL"] != null) {
        cf.setAvatarURL(c["AvatarURL"]);
        }
      }
      thread.headerView.getDiv().style.opacity = "0.0";
      thread.headerView.getDiv().style.top = "${(threads.length-1)*85+60}px";
      headerView.append(thread.headerView.getDiv());
      animate(thread.headerView.getDiv(), duration: 250, properties: {
        'opacity': 1.0
      });

      updateThreadPositions();
    }
  }
  void removeThread(int threadID) {
    ThreadInfo thread = getThread(threadID);
    if (thread != null) {
      threads.remove(thread);
      thread.headerView.getDiv().style.opacity = "1.0";
      animate(thread.headerView.getDiv(), duration: 250, properties: {
        'opacity': 0.0
      }).onComplete.listen((_) {
        thread.headerView.getDiv().remove();
      });
      thread.fullView.getDiv().style.opacity = "1.0";
      animate(thread.fullView.getDiv(), duration: 250, properties: {
        'opacity': 0.0
      }).onComplete.listen((_) {
        thread.fullView.getDiv().remove();
      });
      updateThreadPositions();
      if (thread == viewingThread) {
        restoreToDefaultState();
      }
    }
  }
  void updateThread(int threadID, String title, String description) {
    ThreadInfo thread = getThread(threadID);
    if (thread != null) {
      thread.setTitle(title);
      thread.setDescription(description);
    }
  }
  void moveThreadToTop(int threadID) {
    ThreadInfo thread = getThread(threadID);
    if (thread != null) {
      threads.remove(thread);
      threads.insert(0, thread);
      updateThreadPositions();
    }
  }
  void addComment(Map commentMap) {
    int threadID = commentMap["ThreadID"];
    ThreadInfo thread = getThread(threadID);
    if (thread != null) {
      CommentInfo cf = new CommentInfo(thread, this);
      thread.addComment(cf);
      cf.setThreadID(commentMap['ThreadID']);
      cf.setCommentID(commentMap['CommentID']);
      cf.setComment(commentMap['Comment']);
      cf.setOwner(commentMap['Owner']);
      cf.setAvatarURL(commentMap["AvatarURL"]);
    }
  }
  void deleteComment(int threadID, int commentID) {
    ThreadInfo thread = getThread(threadID);
    if (thread != null) {
      thread.removeComment(commentID);
    }
  }
  void updateComment(int threadID, int commentID, String comment) {
    ThreadInfo thread = getThread(threadID);
    if (thread != null) {
      CommentInfo ci = thread.getComment(commentID);
      ci.setComment(comment);
    }
  }
  //GUI
  void makeBackButtonComponent() {
    //Add the new post button to it
    backButton = new ButtonElement();
    backButton.style
        ..width = "60px"
        ..height = "30px"
        ..backgroundColor = "white"
        ..borderRadius = "4px"
        ..borderStyle = "solid"
        ..borderWidth = "1px"
        ..borderColor = "#DDDDDD"
        ..fontSize = "15px"
        ..top = "5px"
        ..position = "absolute"
        ..outline = "none"
        ..left = "5px";
    backButton.dataset['active'] = "background: #BBBBBB;";
    backButton.text = 'Back';
    backButton.onClick.listen((var d) {
      backButtonClicked();
    });
  }
  void makeNewPostButtonComponent() {
    //Add the new post button to it
    newPostButton = new ButtonElement();
    newPostButton.style
        ..width = "140px"
        ..height = "40px"
        ..backgroundColor = "white"
        ..borderRadius = "4px"
        ..borderStyle = "solid"
        ..borderWidth = "1px"
        ..borderColor = "#DDDDDD"
        ..fontSize = "15px"
        ..top = "5px"
        ..position = "absolute"
        ..outline = "none"
        ..right = "5px";
    newPostButton.dataset['active'] = "background: #BBBBBB;";
    newPostButton.text = 'Create New Post';
    newPostButton.onClick.listen((var d) {
      sectionController.newPostButtonClicked();
    });
  }
}

File: ROID-Forum-Web-Client/src/Interface/CodingSection/ThreadController.ts
import {ThreadInfo} from "./ThreadInfo";
import {Section} from "./Section";
import {Interface} from "../../Utility/Interface";
import {CommentInfo} from "./CommentInfo";
import {Utility} from "../../Utility/Utility";

export class ThreadController {
  threads : ThreadInfo[];
  sectionController : Section;
  mainView : HTMLDivElement;
  headerView : HTMLDivElement;
  fullView : HTMLDivElement;
  newPostButton : HTMLButtonElement;
  backButton : HTMLButtonElement;
  viewingPosition = 0;
  viewingThread : ThreadInfo = null;
  constructor(c : Section) {
    this.sectionController = c;
    this.threads = [];
    this.mainView = Interface.Create({type: 'div'});
    this.headerView = Interface.Create({type: 'div'});
    let s : HTMLSpanElement = Interface.Create({type: 'span'});
    s.innerText = "The Programmer's Corner";
    s.style.position = "absolute";
    s.style.width = "100%";
    s.style.textAlign = "center";
    s.style.top = "14px";
    s.style.fontSize = "25px";
    s.style.color = "white";
    this.headerView.appendChild(s);
    this.fullView = Interface.Create({type: 'div'});
    this.makeNewPostButtonComponent();
    this.makeBackButtonComponent();
    this.mainView.appendChild(this.headerView);
  }
  loggedInEvent() {
    this.headerView.appendChild(this.newPostButton);
    for (let i = 0; i < this.threads.length; i++) {
      let t : ThreadInfo = this.threads[i];
      t.fullView.main.appendChild(t.fullView.commentButtonAndBox);
    }
  }
  loggedOutEvent() {
    this.newPostButton.remove();
    for (let i = 0; i < this.threads.length; i++) {
      let t : ThreadInfo = this.threads[i];
      t.fullView.commentButtonAndBox.remove();
    }
  }
  restoreToDefaultState() {
    this.fullView.remove();
    Utility.ClearElements(this.fullView);
    this.mainView.appendChild(this.headerView);
    this.backButton.remove();
    this.viewingThread = null;
  }
  threadClicked(thread : ThreadInfo) {
    this.headerView.remove();
    this.mainView.appendChild(this.fullView);
    this.fullView.appendChild(thread.fullView.getDiv());
    this.mainView.appendChild(this.backButton);
    this.viewingThread = thread;
  }
  backButtonClicked() {
    this.restoreToDefaultState();
  }
  updateThreadPositions() {
    for (let i = 0; i < this.threads.length; i++) {
      let t : ThreadInfo = this.threads[i];
      t.headerView.getDiv().style.top = `${i*85+60}px`;
      //animate(t.headerView.getDiv(), duration: 250, properties: {
      //  'top': "${i*85+60}px"
      //});
    }
  }
  getThread(id : number) : ThreadInfo {
    for (let t of this.threads) {
      if (t.getID() == id) return t;
    }
    return null;
  }
  showThread(threadID : number) {
    let thread : ThreadInfo = this.getThread(threadID);
    if (thread != null) {
      this.threadClicked(thread);
    }
  }
  //Actions
  clearAllThreads() {
    for (let i = 0; i < this.threads.length; i++) {
      let t : ThreadInfo = this.threads[i];
      t.headerView.getDiv().remove();
      t.fullView.getDiv().remove();
    }
    this.threads = [];
  }
  addThread(threadMap : any) {
    let id : number = threadMap["ID"];
    if (this.getThread(id) == null) {
      let thread : ThreadInfo = new ThreadInfo(this);
      thread.setID(id);
      this.threads.push(thread);

      thread.setOwner(threadMap["Owner"]);
      if (threadMap["AvatarURL"] != null) {
        thread.setAvatarURL(threadMap["AvatarURL"]);
      }
      thread.setTitle(threadMap["Title"]);
      thread.setDescription(threadMap["Description"]);

      let commentArray : any[] = threadMap["Comments"];
      for (let i = 0; i < commentArray.length; i++) {
        let c : any = commentArray[i];
        let cf : CommentInfo = new CommentInfo(thread, this);
        thread.addComment(cf);
        console.log(`Loaded comment with threadID: ${c['ThreadID']} and commentID: ${c['CommentID']}`);
        cf.setThreadID(c['ThreadID']);
        cf.setCommentID(c['CommentID']);
        cf.setComment(c['Comment']);
        cf.setOwner(c['Owner']);
        if (c["AvatarURL"] != null) {
        cf.setAvatarURL(c["AvatarURL"]);
        }
      }
      thread.headerView.getDiv().style.opacity = "1.0";
      thread.headerView.getDiv().style.top = `${(this.threads.length-1)*85+60}px`;
      this.headerView.appendChild(thread.headerView.getDiv());
      //animate(thread.headerView.getDiv(), duration: 250, properties: {
      //  'opacity': 1.0
      //});

      this.updateThreadPositions();
    }
  }
  removeThread(threadID : number) {
    let thread : ThreadInfo = this.getThread(threadID);
    if (thread != null) {
      this.threads.splice(this.threads.indexOf(thread), 1);
      thread.headerView.getDiv().style.opacity = "1.0";
      //animate(thread.headerView.getDiv(), duration: 250, properties: {
      //  'opacity': 0.0
      //}).onComplete.listen((_) {
        thread.headerView.getDiv().remove();
      //});
      thread.fullView.getDiv().style.opacity = "1.0";
      //animate(thread.fullView.getDiv(), duration: 250, properties: {
      //  'opacity': 0.0
      //}).onComplete.listen((_) {
        thread.fullView.getDiv().remove();
      //});
      this.updateThreadPositions();
      if (thread == this.viewingThread) {
        this.restoreToDefaultState();
      }
    }
  }
  updateThread(threadID : number, title : string, description : string) {
    let thread : ThreadInfo = this.getThread(threadID);
    if (thread != null) {
      thread.setTitle(title);
      thread.setDescription(description);
    }
  }
  moveThreadToTop(threadID : number) {
    let thread : ThreadInfo = this.getThread(threadID);
    if (thread != null) {
      this.threads.splice(this.threads.indexOf(thread), 1);
      this.threads.splice(0, 0, thread);
      this.updateThreadPositions();
    }
  }
  addComment(commentMap : any) {
    let threadID : number = commentMap["ThreadID"];
    let thread : ThreadInfo = this.getThread(threadID);
    if (thread != null) {
      let cf : CommentInfo = new CommentInfo(thread, this);
      thread.addComment(cf);
      cf.setThreadID(commentMap['ThreadID']);
      cf.setCommentID(commentMap['CommentID']);
      cf.setComment(commentMap['Comment']);
      cf.setOwner(commentMap['Owner']);
      cf.setAvatarURL(commentMap["AvatarURL"]);
    }
  }
  deleteComment(threadID : number, commentID : number) {
    let thread : ThreadInfo = this.getThread(threadID);
    if (thread != null) {
      thread.removeComment(commentID);
    }
  }
  updateComment(threadID : number, commentID : number, comment : string) {
    let thread : ThreadInfo = this.getThread(threadID);
    if (thread != null) {
      let ci : CommentInfo = thread.getComment(commentID);
      ci.setComment(comment);
    }
  }
  //GUI
  makeBackButtonComponent() {
    //Add the new post button to it
    this.backButton = Interface.Create({type: 'button'});
        this.backButton.style.width = "60px";
        this.backButton.style.height = "30px";
        this.backButton.style.backgroundColor = "white";
        this.backButton.style.borderRadius = "4px";
        this.backButton.style.borderStyle = "solid";
        this.backButton.style.borderWidth = "1px";
        this.backButton.style.borderColor = "#DDDDDD";
        this.backButton.style.fontSize = "15px";
        this.backButton.style.top = "5px";
        this.backButton.style.position = "absolute";
        this.backButton.style.outline = "none";
        this.backButton.style.left = "5px";
    this.backButton.dataset['active'] = "background: #BBBBBB;";
    this.backButton.innerText = 'Back';
    this.backButton.onclick = () => {
      this.backButtonClicked();
    };
  }
  makeNewPostButtonComponent() {
    //Add the new post button to it
    this.newPostButton = Interface.Create({type: 'button'});
        this.newPostButton.style.width = "140px";
        this.newPostButton.style.height = "40px";
        this.newPostButton.style.backgroundColor = "white";
        this.newPostButton.style.borderRadius = "4px";
        this.newPostButton.style.borderStyle = "solid";
        this.newPostButton.style.borderWidth = "1px";
        this.newPostButton.style.borderColor = "#DDDDDD";
        this.newPostButton.style.fontSize = "15px";
        this.newPostButton.style.top = "5px";
        this.newPostButton.style.position = "absolute";
        this.newPostButton.style.outline = "none";
        this.newPostButton.style.right = "5px";
    this.newPostButton.dataset['active'] = "background: #BBBBBB;";
    this.newPostButton.innerText = 'Create New Post';
    this.newPostButton.onclick = () => {
      this.sectionController.newPostButtonClicked();
    };
  }
}
